x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/multiterms/MultiTermsAggregator.java

```diff
@@ -0,0 +1,538 @@
+     */
+    static List<Object> unpackTerms(BytesRef termsBytes) {
+        try (StreamInput input = new BytesArray(termsBytes).streamInput()) {
+            return input.readList(StreamInput::readGenericValue);
+        } catch (IOException ex) {
+            throw ExceptionsHelper.convertToRuntime(ex);
+        }
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        List<TermValues> termValuesList = termValuesList(ctx);
+
+        return new LeafBucketCollectorBase(sub, values) {
+            @Override
+            public void collect(int doc, long owningBucketOrd) throws IOException {
+                List<List<Object>> terms = docTerms(termValuesList, doc);
+                if (terms != null) {
+                    termsIterator(terms, owningBucketOrd, bucketOrd -> {
+                        if (bucketOrd < 0) { // already seen
+                            bucketOrd = -1 - bucketOrd;
+                            collectExistingBucket(sub, doc, bucketOrd);
+                        } else {
+                            collectBucket(sub, doc, bucketOrd);
+                        }
+                    });
+                }
+            }
+        };
+    }
+
+    /**
+     * Iterates through a list of terms and generates an n-ary Cartesian product of all terms that is passed to consumer.
+     *
+     * This is needed so we can generate all permutations.
+     */
+    private void termsIterator(List<List<Object>> terms, long owningBucketOrd, CheckedConsumer<Long, IOException> consumer)
+        throws IOException {
+        List<Object> path = new ArrayList<>(terms.size());
+        new CheckedConsumer<Integer, IOException>() {
+            @Override
+            public void accept(Integer start) throws IOException {
+                for (Object term : terms.get(start)) {
+                    if (start == path.size()) {
+                        path.add(term);
+                    } else {
+                        path.set(start, term);
+                    }
+                    if (start < terms.size() - 1) {
+                        this.accept(start + 1);
+                    } else {
+                        long bucketOrd = bucketOrds.add(owningBucketOrd, packKey(path));
+                        consumer.accept(bucketOrd);
+                    }
+                }
+            }
+        }.accept(0);
+    }
+
+    @Override
+    public InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws IOException {
+        InternalMultiTerms.Bucket[][] topBucketsPerOrd = new InternalMultiTerms.Bucket[owningBucketOrds.length][];
+        long[] otherDocCounts = new long[owningBucketOrds.length];
+        for (int ordIdx = 0; ordIdx < owningBucketOrds.length; ordIdx++) {
+            collectZeroDocEntriesIfNeeded(owningBucketOrds[ordIdx]);
```

Comment:
I wonder if we should just remove this functionality for multi-terms?  It's always been strange functionality in terms imo, and I could see it being very expensive for multi-terms if there are a lot of high cardinality fields?

Reply:
Hmm, good point. We should probably remove it.